include/unifex/async_mutex.hpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/detail/atomic_intrusive_queue.hpp>
+#include <unifex/detail/intrusive_queue.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+
+namespace unifex {
+
+class async_mutex {
+  class lock_sender;
+
+public:
+  async_mutex() noexcept;
+  async_mutex(const async_mutex &) = delete;
+  async_mutex(async_mutex &&) = delete;
+  ~async_mutex();
+
+  async_mutex &operator=(const async_mutex &) = delete;
+  async_mutex &operator=(async_mutex &&) = delete;
+
+  [[nodiscard]] bool try_lock() noexcept;
+
+  [[nodiscard]] lock_sender async_lock() noexcept;
+
+  void unlock() noexcept;
+
+private:
+  struct waiter_base {
+    void (*resume_)(waiter_base *) noexcept;
+    waiter_base *next_;
+  };
+
+  class lock_sender {
+  public:
+    template <template <typename...> class Variant,
+              template <typename...> class Tuple>
+    using value_types = Variant<Tuple<>>;
+
+    template <template <typename...> class Variant>
+    using error_types = Variant<>;
+
+  private:
+    friend async_mutex;
+
+    explicit lock_sender(async_mutex &mutex) noexcept : mutex_(mutex) {}
+
+    lock_sender(const lock_sender &) = delete;
+    lock_sender(lock_sender &&) = default;
+
+    template <typename Receiver> class operation : waiter_base {
+      friend lock_sender;
+
+      template <typename Receiver2>
+      explicit operation(async_mutex &mutex, Receiver2 &&r) noexcept
+          : mutex_(mutex), receiver_((Receiver2 &&) r) {
+        this->resume_ = [](waiter_base * self) noexcept {
+          operation &op = *static_cast<operation *>(self);
+          cpo::set_value((Receiver &&) op.receiver_);
+        };
+      }
+
+      operation(operation &&) = delete;
+
+      friend void tag_invoke(tag_t<cpo::start>, operation &op) noexcept {
+        if (!op.mutex_.try_enqueue(&op)) {
+          // Failed to enqueue because we acquired the lock
+          // synchronously. Invoke the continuation inline
+          // without type-erasure here.
+          cpo::set_value((Receiver &&) op.receiver_);
```

Comment:
Can invoking the continuation inline here cause problems with deep recursion?

Reply:
Yes, this can be a problem.

I was planning to make a subsequent change to have async_mutex require a 'current scheduler' to be provided by the receiver and to use it to schedule a call to the receiver in the case that the operation completes asynchronously (ie. don't call the next lock acquirer inline within the unlock() method).

Unbounded recursion is a general problem, however, and any code using iteration will need to explicitly inject some kind of schedule operation, either with the ambient scheduler, or with a trampoline_scheduler.